Added experimental vicuna 1.1+ support

jphme opened the pull request:
As discussed in #262 and #261 , I implemented support for Vicuna 1.1+, following the [original code](https://github.com/lm-sys/FastChat/blob/main/fastchat/train/train.py) as close as possible.

This should be even more flexible than Vicunas own finetuning/training code, as my code allows for flexible system messages.

Dataset should look like 
```
[{"conversation":{"from": "human", "value": "Who are you?"}, {"from": "gpt", "value": "I am Vicuna"},...}]
```
as .jsonl file; the first "from" can be "system" instead and will be used as system message if present.

Submitting as draft as the code is still untested, but will report back.

Commit:
added experimental vicuna 1.1+ support

Commit:
small fix

Commit:
fix

Commit:
fix tensor to list

Nan-Do commented:
I just finished the simple-instruct version and I was going to send it. Luckily I saw this before creating the PR. Waiting to see the results

Commit:
fixed a few small errors

Commit:
small fix with truncated prompts

Commit:
small fix for truncated conversations

Commit:
vicuna padding happens earlier, fix for removal

jphme commented:
Everything should work now, a large finetuning of Vicuna 13b v.1.3 with this implementation is currently in progress and I am excited for the results 💯 .

Commit:
pre-commits passing

jphme commented:
I finetuned Vicuna 1.3 13B sucessfully with this implementation so should be ready for merging.

There is some duplicate code and/or some duplicate functionality, but as I understood, the plan is to make `prompt_strategies` modular / independent of each other and thus an implementation like this one, that is very close to the original refence implementation and more or less self-contained should be in line with this idea?

valeman100 commented:
Hello, thank you for your implementation of Vicuna dataset processing.

May I ask you to specify which is the correct dataset type to use?
In your implementation, I found:
"Use dataset type: orcamini in config.yml to use this prompt style."

but I can't find the orcamini config anywhere.

Commit:
fixed typo and instructions

jphme commented:
> Hello, thank you for your implementation of Vicuna dataset processing.
> 
> May I ask you to specify which is the correct dataset type to use? In your implementation, I found: "Use dataset type: orcamini in config.yml to use this prompt style."
> 
> but I can't find the orcamini config anywhere.

@valeman100 ahh, that was an oversight (copy&pasted from my orcanmini implementation), thanks for the heads up, fixed it and added another example for the configuration in the config.yml to make it more explicit.

NanoCode012 commented on src/axolotl/prompt_strategies/vicuna.py:
````diff
@@ -0,0 +1,176 @@
+
+The dataset itself should look like this:
+```
+{'conversation':[{"from": "human", "value": "Who are you?"}, {"from": "gpt", "value": "I am Vicuna"},...]}
````
I'm not so clear on this. Should it be `conversations` with `s`? This will allow it to be compatible with older sharegpt datasets.

NanoCode012 commented:
One note, there was a recent PR to improve tokenize speed. However, when I apply your branch to the current main with a normal sharegpt dataset, I would get the below error. I am unclear whether it's due to that PR or some mismatch with yours.

<details>

<summary>error message log</summary>

```
Traceback (most recent call last):
  File "/workspace/axolotl/scripts/finetune.py", line 356, in <module>
    fire.Fire(train)
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/workspace/axolotl/scripts/finetune.py", line 226, in train
    train_dataset, eval_dataset = load_prepare_datasets(
  File "/workspace/axolotl/src/axolotl/utils/data.py", line 352, in load_prepare_datasets
    dataset = load_tokenized_prepared_datasets(
  File "/workspace/axolotl/src/axolotl/utils/data.py", line 268, in load_tokenized_prepared_datasets
    samples = samples + list(d)
  File "/workspace/axolotl/src/axolotl/datasets.py", line 41, in __iter__
    self.dataset.map(
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/datasets/arrow_dataset.py", line 580, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/datasets/arrow_dataset.py", line 545, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/datasets/arrow_dataset.py", line 3087, in map
    for rank, done, content in Dataset._map_single(**dataset_kwargs):
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/datasets/arrow_dataset.py", line 3441, in _map_single
    example = apply_function_on_filtered_inputs(example, i, offset=offset)
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/datasets/arrow_dataset.py", line 3344, in apply_function_on_filtered_inputs
    processed_inputs = function(*fn_args, *additional_args, **fn_kwargs)
  File "/workspace/axolotl/src/axolotl/prompt_strategies/vicuna.py", line 78, in tokenize_prompt
    input_ids = self.tokenizer(
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 2577, in __call__
    encodings = self._call_one(text=text, text_pair=text_pair, **all_kwargs)
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 2683, in _call_one
    return self.encode_plus(
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 2756, in encode_plus
    return self._encode_plus(
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils.py", line 652, in _encode_plus
    return self.prepare_for_model(
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 3235, in prepare_for_model
    encoded_inputs = self.pad(
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 3040, in pad
    encoded_inputs = self._pad(
  File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 3425, in _pad
    encoded_inputs["attention_mask"] = encoded_inputs["attention_mask"] + [0] * difference
OverflowError: cannot fit 'int' into an index-sized integer
```

</details>

jphme commented:
> One note, there was a recent PR to improve tokenize speed. However, when I apply your branch to the current main with a normal sharegpt dataset, I would get the below error. I am unclear whether it's due to that PR or some mismatch with yours.
> error message log
> 
> ```
> Traceback (most recent call last):
>   File "/workspace/axolotl/scripts/finetune.py", line 356, in <module>
>     fire.Fire(train)
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/fire/core.py", line 141, in Fire
>     component_trace = _Fire(component, args, parsed_flag_args, context, name)
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/fire/core.py", line 475, in _Fire
>     component, remaining_args = _CallAndUpdateTrace(
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
>     component = fn(*varargs, **kwargs)
>   File "/workspace/axolotl/scripts/finetune.py", line 226, in train
>     train_dataset, eval_dataset = load_prepare_datasets(
>   File "/workspace/axolotl/src/axolotl/utils/data.py", line 352, in load_prepare_datasets
>     dataset = load_tokenized_prepared_datasets(
>   File "/workspace/axolotl/src/axolotl/utils/data.py", line 268, in load_tokenized_prepared_datasets
>     samples = samples + list(d)
>   File "/workspace/axolotl/src/axolotl/datasets.py", line 41, in __iter__
>     self.dataset.map(
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/datasets/arrow_dataset.py", line 580, in wrapper
>     out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/datasets/arrow_dataset.py", line 545, in wrapper
>     out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/datasets/arrow_dataset.py", line 3087, in map
>     for rank, done, content in Dataset._map_single(**dataset_kwargs):
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/datasets/arrow_dataset.py", line 3441, in _map_single
>     example = apply_function_on_filtered_inputs(example, i, offset=offset)
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/datasets/arrow_dataset.py", line 3344, in apply_function_on_filtered_inputs
>     processed_inputs = function(*fn_args, *additional_args, **fn_kwargs)
>   File "/workspace/axolotl/src/axolotl/prompt_strategies/vicuna.py", line 78, in tokenize_prompt
>     input_ids = self.tokenizer(
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 2577, in __call__
>     encodings = self._call_one(text=text, text_pair=text_pair, **all_kwargs)
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 2683, in _call_one
>     return self.encode_plus(
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 2756, in encode_plus
>     return self._encode_plus(
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils.py", line 652, in _encode_plus
>     return self.prepare_for_model(
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 3235, in prepare_for_model
>     encoded_inputs = self.pad(
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 3040, in pad
>     encoded_inputs = self._pad(
>   File "/workspace/axolotl/venv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 3425, in _pad
>     encoded_inputs["attention_mask"] = encoded_inputs["attention_mask"] + [0] * difference
> OverflowError: cannot fit 'int' into an index-sized integer
> ```

Hm strange, will test this out again (and also double check tokenization).

Can you put this PR in draft mode again? I am currently working on llama2 chat and will try to fix this afterwards...

NanoCode012 commented on src/axolotl/prompt_strategies/vicuna.py:
```diff
@@ -0,0 +1,176 @@
+            conversation_str,
+            return_tensors="pt",
+            padding="max_length",
+            max_length=self.tokenizer.model_max_length,
```
```suggestion
            max_length=self.sequence_len,
```

This is problematic for llamav2 which has some randomly high number. Please let me know if this breaks any intended feature. Following other tokenizing code here, we need also set it follow seq len.

```json
"model_max_length": 1000000000000000019884624838656,
```

NanoCode012 commented on src/axolotl/prompt_strategies/vicuna.py:
```diff
@@ -0,0 +1,176 @@
+
+        target[cur_len:] = IGNORE_TOKEN_ID
+
+        if cur_len < self.tokenizer.model_max_length:
```
```suggestion
        if cur_len < self.sequence_len:
```

If you accept the above changes, please accept this too.

NanoCode012 commented:
Making those modifications fixed it. However, I'm getting a few tokenization mismatch, a lot being cur_len = total_len + 1. Is this normal?

valeman100 commented:
> Making those modifications fixed it. However, I'm getting a few tokenization mismatch, a lot being cur_len = total_len + 1. Is this normal?

I found out the same, try to set use_fast=False in the tokenizer and see if it is related to the fast tokenizer

NanoCode012 commented:
@valeman100 , thanks for the response. No, it didn't. I'm testing this with llama-2-7b-hf with convo sharegpt dataset for reference.

jphme committed:
Update src/axolotl/prompt_strategies/vicuna.py

Co-authored-by: NanoCode012 <kevinvong@rocketmail.com>

jphme committed:
Update src/axolotl/prompt_strategies/vicuna.py

Co-authored-by: NanoCode012 <kevinvong@rocketmail.com>

jphme commented:
> @valeman100 , thanks for the response. No, it didn't. I'm testing this with llama-2-7b-hf with convo sharegpt dataset for reference.

I will check this again if Llama2 is running. 

Just fyi, yesterday I almost got crazy testing #296  because I always got exactly these tokenization mismatches by one when running a test-finetune but not when testing locally.

This was due to my config.yml which included:
```
special_tokens:
  bos_token: "<s>"
  eos_token: "</s>"
  unk_token: "<unk>"
  ```

removing these everyting works fine.

These were also included in my config during my successful Vicuna finetune, so could be related to that.

NanoCode012 commented:
I added a pad_token [PAD] as the llama2 doc states and ref here #294 .

Also, my eval turns to NAN... Could be due to all those misses..

Edit: `sharegpt:chat` works for me though.

jphme commented:
> I added a pad_token [PAD] as the llama2 doc states and ref here #294 .
> 
> Also, my eval turns to NAN... Could be due to all those misses..
> 
> Edit: `sharegpt:chat` works for me though.

I am bit confused, did you mean #296 ? As there, I added the pad-token manually, but you are right that you would need to include it elswhere excep in the prompting_strategy itself.

I will check Vicuna tokenization and will try to bring this in a clean, ready-to-merge state with test fixtures at beginning of next week.

flexchar commented:
have you had a chance to look at this @jphme?

NanoCode012 commented:
@flexchar , sorry if this is late. We have imported `fastchat` library, so you can now use the same as fastchat's convo.

You can check the docs
```yaml
datasets:
  - path: 
    type: sharegpt
    conversation:  # Optional[str] fastchat conversation type, only used with type: sharegpt
 ```

Vicuna 1.1 should be just `sharegpt`, but you can even use other's by modifying conversation.

Nan-Do commented:
I checked this back in the day, for the llama2_chat code, and the problem with the mismatch of the tokens happens because the tokenizer was initialized incorrectly. See the following https://github.com/huggingface/transformers/pull/24565. If you are adding new tokens during the retraining you need to be aware of this and set the flag properly.  Axolotl can already set this flag https://github.com/OpenAccess-AI-Collective/axolotl/issues/305. (The llama2_chat has some other issues perhaps I should open another issue to address them)

flexchar commented:
Thank you for following up @NanoCode012. I really appreciate that. It's still interesting to try but these days I'm moving onto Mistral and my preferred flavour of model (MistralOrca) uses ChatML template. I see `alpaca_w_system.load_open_orca_chatml` is now available for that. :)

jphme commented:
> Thank you for following up @NanoCode012. I really appreciate that. It's still interesting to try but these days I'm moving onto Mistral and my preferred flavour of model (MistralOrca) uses ChatML template. I see `alpaca_w_system.load_open_orca_chatml` is now available for that. :)

For conversation datasets, you actually should be able to do:

```
datasets:
  - path: 
    type: sharegpt
    conversation: chatml
```

if I´m not mistaken, see https://github.com/OpenAccess-AI-Collective/axolotl/blob/490923fb78e0646c2c0ba427628a0daa72b86996/src/axolotl/prompt_strategies/sharegpt.py#L11 . 

Sorry I totally forgot about this PR and think it can be closed/deleted for good :) .
